hathor/builder.py

```diff
@@ -36,11 +40,266 @@
 from hathor.p2p.utils import discover_hostname
 from hathor.prometheus import PrometheusMetricsExporter
 from hathor.pubsub import PubSubManager
+from hathor.storage import RocksDBStorage
+from hathor.transaction.storage import TransactionMemoryStorage, TransactionRocksDBStorage, TransactionStorage
+from hathor.util import Random, Reactor, get_environment_info
 from hathor.wallet import BaseWallet, HDWallet, Wallet
 
 logger = get_logger()
 
 
+class StorageType(Enum):
+    MEMORY = 'memory'
+    ROCKSDB = 'rocksdb'
+
+
+class BuildArtifacts(NamedTuple):
+    settings: HathorSettingsType
+    rng: Random
+    reactor: Reactor
+    manager: HathorManager
+    pubsub: PubSubManager
+    consensus: ConsensusAlgorithm
+    tx_storage: TransactionStorage
+    indexes: Optional[IndexesManager]
+    wallet: Optional[BaseWallet]
+    rocksdb_storage: Optional[RocksDBStorage]
+
+
+class Builder:
+    def __init__(self) -> None:
+        self.log = logger.new()
+        self.artifacts: Optional[BuildArtifacts] = None
+
+        self._rng: Random = Random()
+        self._checkpoints: Optional[List[Checkpoint]] = None
+        self._capabilities: Optional[List[str]] = None
+
+        self._peer_id: Optional[PeerId] = None
+        self._network: Optional[str] = None
+        self._cmdline: str = ''
+
+        self._storage_type: Optional[StorageType] = None
+        self._force_memory_index: bool = False
+
+        self._event_manager: Optional[EventManager] = None
+
+        self._rocksdb_path: Optional[str] = None
+        self._rocksdb_storage: Optional[RocksDBStorage] = None
+
+        self._tx_storage: Optional[TransactionStorage] = None
+
+        self._reactor: Optional[Reactor] = None
+        self._pubsub: Optional[PubSubManager] = None
+
+        self._wallet: Optional[BaseWallet] = None
+        self._wallet_directory: Optional[str] = None
+        self._wallet_unlock: Optional[bytes] = None
+
+        self._enable_address_index: bool = False
+        self._enable_tokens_index: bool = False
+        self._enable_utxo_index: bool = False
+
+        self._enable_sync_v1: Optional[bool] = None
+        self._enable_sync_v2: Optional[bool] = None
+
+        self._stratum_port: Optional[int] = None
+
+    def build(self) -> BuildArtifacts:
+        if self.artifacts is not None:
+            return self.artifacts
```

Comment:
Is there any application currently for being able to call `builder.build()` multiple times? Otherwise I'd rather have it only build once and fail on next calls, and if the user of `builder` needs it can persist `build_artifacts` instead of the `builder`.

Otherwise I think it opens up to this kind of silent error:

```python
builder = Builder()
builder.set_rng(...)  # or any other set method
... = builder.build()
# this can happen on another interface
builder.set_rng(...)  # this will silently fail, that is it will not have any effect on the next call
... = builder.build()
```

Maybe it's worth to block any `set_*` method after `build` is called (in Rust for example this would be prevented by making the `build` method "consume" the builder object and prevent it from being used after that, I'm not sure if it's possible to make `del self` inside of `build` right before returning, which would have a similar effect). Either way, IMO it would be enough to just `raise ValueError('cannot call build twice')` here and leave other improvements for the future.

What do you think?

Reply:
Agreed. Done!